proposals/subclass-opt-in-required.md

```diff
@@ -0,0 +1,131 @@
+  because it may require 3rd-party authors to re-compile their code in order to be compatible with the original API.
+- **Stable to use but closed for 3rd-party implementations.** Typically it is an API that is supposed to be sealed, 
+  but cannot for technical reasons (e.g. being spread over multiple modules). In such scenarios, authors may
+  rely on implementation details of their API (i.e. downcasting for performance reasons), but use-sites of APIs are stable.
+  - It also includes API that can be technically sealed, but due to limitations of the language, authors would like to let it be
+    open in order to avoid future compatibility issues with their own extensions of the API and 3rd-party exhaustiveness checks.
+- **Stable to use but delicate to implement.** It is an API with very non-trivial API contracts, where individual methods
+  (potentially over multiple interfaces) should behave in a coordinated manner or where it is known that implementation by delegation
+   produces [unwanted result](https://youtrack.jetbrains.com/issue/KT-18324).
+- **Stable to use but with a contract that may be weakened in the future**. For example, an API may change its _input_ parameter type `T` to `T?`.
+  Such change won't break existing users as they previously had no means to pass `null`, but will break existing implementations that did not expect `null`. 
+
+These patterns imply that **most of the uses** in the wild are stable and the cost of marking original declarations with `OptIn` 
+will create unnecessary inconvenience for such uses. On the other hand, leaving the API as is will create a false sense of
+security for users and library authors that extend it.
+
+## Goals
+
+* Clear **implementation opt-in** is required for all users whose code may be broken by changes in the API implementation contract.
+* **Consistency** with existing `OptIn` API in a way that users already familiar with `OptIn` will also be familiar with the proposed solution.
+* **Robustness**. It should be possible to apply the solution only to extensible parts of API.
+* **Smooth graduation**: if the API doesn't need to be changed as soon as it *graduates* (i.e. no longer requires opt-in), just unmark it, and all clients remain binary compatible.
+* **Garbage collection upon graduation**: when the API graduates, the opt-in flags/markers are highlighted as warnings (and subsequently, errors) so that clients know to remove them.
+
+
+## Proposal
+
+We propose to add the following declaration to the standard library:
+
+```kotlin
+package kotlin
+
+@Target(CLASS)
+@Retention(BINARY)
+@SinceKotlin("1.8")
+@ExperimentalSubclassOptIn
+public annotation class SubclassOptInRequired(
+    val markerClass: KClass<out Annotation>
+)
+```
+
+The `SubclassOptInRequired` accepts only opt-in markers (annotations that are explicitly marked with `RequiresOptIn`) and can be 
```

Comment:
The name makes the impression that only deriving a class (subclass) from this type requires an opt-in, whereas it affects derived interfaces as well.

Reply:
This name is explained in follow-up `Naming alternatives` section.
We failed to come up with a much better name

Reply:
We are going to branch-out 1.8.0 soon and feature is available as experimental (guarded by opt-in), so if you have any objections or suggestions, it's time to raise the question to our DM

Reply:
Did you consider subtype?

Reply:
Subtyping is considered to be a use-site beast while subclassing is the declaration-site, and this feature is purely declaration-site based. Also, subtyping is a more overloaded term as it implies generics and variance binary relation